### ISSUE TITLE
Forms Demo V2

### ISSUE THREAD

**Issue by rachael-phillips** [patternfly/patternfly-next#919]
Continued revision on Forms Demo based on requirements from #389

**Comment by rachael-phillips** [comment]
@mcarrano Does @mcoker need to consider this while working on #457 ?

**Comment by mcarrano** [comment]
I don't think so. But do you see a connection that I may be missing @rachael-phillips ?

**Comment by stacymcauliffe** [comment]
As-is, we have three different types of help:
-Helper text: Content below a form that gives the user additional critical info about the input they need to provide, e.g. Project name must be unique
-Field level help: popover content that novice users would find useful when filling out a form, e.g. CVV is the 3-digit number on the back of the credit card.
-Syntax hint: Give the user info about the specific format they need to provide data in, e.g. MM/DD/YYYY

Issues
-Field level help popover is currently placed to the right of the input field which doesn't work for a11y. 
-Syntax hints are provided on the top of the input field which might make things feel crowded.
- Helper text is provided underneath the input field, but why should we treat helper text differently from syntax hints

Resolutions from this meeting to try for next time
-Move the popover to the top and visually associate with label 
-Move syntax hint into the input field - precede with e.g. to account for a11y contrast issues. Formatting should be as follows: e.g. John Smith , e.g. jon@smith.com, e.g. (781) 555-5555
Any errors encountered by the user as a result of an input error will restate the syntax guidance inline beneath the form field. So, if a syntax hint reads: e.g. jon@smith.com, then the error would read, Input a valid email address, e.g. jon @smith.com
-Keep helper text as is

From Matt: The only thing I would add is the labeling can be either to the left or top of the field, so the examples we create for Field Level Help should cover both. Also need to consider placement and spacing of the "?" icon when there is also a required field indicator.

**Comment by mceledonia** [comment]
Some visual updates to forms:

- This example shows optional placeholder text pattern. Recommending against placeholder text altogether, but also providing this solution for users who must have it.
- Moving infotip icon to left of field label
- Added required asterisks 

[Forms.zip](https://github.com/patternfly/patternfly-next/files/2860839/Forms.zip)

@mcarrano @stacymcauliffe

**Comment by jgiardino** [comment]
This seems the Forms epic issue, so I'm adding another issue related to forms:
https://github.com/patternfly/patternfly-next/issues/1420

**Comment by mceledonia** [comment]
Update - closing this one out soon, setting up a meeting with dev + interaction to finalize.

**Comment by mcarrano** [comment]
@mceledonia can you post your latest design mocks here? Also, a question - I like the way we are now constraining the width of a form when it is placed in a page. But I question whether the white background should fill the width of the page (just the background, not the form). Too me it looks odd with the empty gray space to the right of the form, but maybe that's just me. What are your thoughts?

**Comment by mceledonia** [comment]
@mcarrano I could go either way on that, I don't feel too strongly but I can try to explain my reasoning:

Since the form itself has a limited width (in this case) to visually mark that width, the card/background acts as a good indication of this. With the white extending the full length of the page, it felt more like the form width was unintentional because the container it lives in was a different size than the form itself. Again though I see how the extra gray space could have a similar issue, so like I said I could go either way. Maybe we could do some quick testing with others to find a consensus.

**Comment by mcarrano** [comment]
Yeah, I don't really have a good argument for whether one approach is more usable than the other, @mceledonia . I'm content to go with what you've designed and see if there is any further push back about it.

**Comment by rachael-phillips** [comment]
@mceledonia would you mind updating this issue with the finalized designs? :) thanks so much!

**Comment by rachael-phillips** [comment]
@mceledonia would you mind updating this issue with the finalized designs? :) thanks so much!

**Comment by mceledonia** [comment]
Definitely, I'll edit this post and upload after my morning meetings today.

**Comment by mceledonia** [comment]
![Forms - V2 - Widths](https://user-images.githubusercontent.com/30935938/64355552-6cad7e00-cfcf-11e9-8061-86aebc40ca75.png)
![Forms - V2 Left Labels](https://user-images.githubusercontent.com/30935938/64355553-6cad7e00-cfcf-11e9-90d5-8826f8ca5ba6.png)
![Forms - V2](https://user-images.githubusercontent.com/30935938/64355556-6cad7e00-cfcf-11e9-9731-d9d97fa41130.png)
![Max Width Context Example](https://user-images.githubusercontent.com/30935938/64355557-6cad7e00-cfcf-11e9-9cf1-2f0a92b1c396.png)

**Comment by mcarrano** [comment]
One question @mceledonia . These mockups still use the open circle-? for pop-up help. Per @kybaker I understood we were changing to the filled circle-? as updated in the design documentation here: https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/forms  Can we clarify and update if needed?

**Comment by mcarrano** [comment]
Want to make sure that placeholder text in form examples represents best practices. See https://github.com/patternfly/patternfly-org/issues/1614

**Comment by mcarrano** [comment]
Added https://github.com/patternfly/patternfly-next/issues/2827 to this epic to illustrate how text can be added following a text field to reflect a calculated value.

**Comment by mmenestr** [comment]
Just went through this issue to see if it conflicted with the updates we are currently making to the form guidelines and wanted to flag a few things:
1. The "cancel" button should no longer use a secondary button, and should be a link button instead!
2. I can confirm that the final icon decision for popover icons is going to be the **un-filled** question mark circle!
3. This is more of a question but @mceledonia for the third example "in context using the max width" in a full page, should that still be shown with a white card background? I'm asking because in my documentation examples we'd decided to get rid of the white card background and just place it directly on the white page! So I want to make sure it aligns in the react/html examples!

**Comment by jgiardino** [comment]
@mceledonia In this example, where would the help icon display relative to the Address field, just to the right of the text label, or just to the left of the input field? I

![image](https://user-images.githubusercontent.com/21063328/78719540-10f2fc80-78f2-11ea-950b-7959c2ccdab2.png)

**Comment by mmenestr** [comment]
@jgiardino it should be placed to the **right of the text label,** the same way it would be placed in a top-aligned form field!

**Comment by mcarrano** [comment]
Transferring back to design to review and update, if necessary. A clean core issue will be opened when done.
